### PR TITLE
Added Distributed Cache implementation of User Session Store

### PR DESCRIPTION
**Pull request by PascalSenn**
**What issue does this PR address?**
This PR adds a implementation of the user session store building on top of IDistributedCache

Current Limitations
- Only the newest Session can be fetched from the cache. 

It's not possible to store lists on the cache without serialisation. 
This could be implemented, but the question is is it really needed
How often does it happen that there are multiple sessions of the same subject and you want to access the old session?
Is searching by multiple keys (e.g. sessionId, sub) often used?

**Important: Any code or remarks in your Pull Request are under the following terms:**

If You provide us with any comments, bug reports, feedback, enhancements, or modifications proposed or suggested by You for the Software, such Feedback is provided on a non-confidential basis (notwithstanding any notice to the contrary You may include in any accompanying communication), and Licensor shall have the right to use such Feedback at its discretion, including, but not limited to the incorporation of such suggested changes into the Software. You hereby grant Licensor a perpetual, irrevocable, transferable, sublicensable, nonexclusive license under all rights necessary to incorporate and use your Feedback for any purpose, including to make and sell any products and services.

(see [our license](https://duendesoftware.com/license/identityserver.pdf), section 7)

### CODE REVIEW

File: src/Duende.Bff.DistributedCache/DistributedCacheUserSessionStore.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            UserSession? sessionBySubjectId = null;
+            UserSession? sessionBySessionId = null;
+            if (!string.IsNullOrWhiteSpace(filter.SubjectId))
+            {
+                Identifier identifier =
+                    new Identifier(_applicationDiscriminator, filter.SubjectId, Subject);
+
+                sessionBySubjectId =
+                    await _distributedCache.GetUserSessionAsync(identifier, cancellationToken);
+            }
```
Comment (PascalSenn) [inline src/Duende.Bff.DistributedCache/DistributedCacheUserSessionStore.cs:129]: This is a limitation at the moment. 
We can only get the most recent session 

Given you store the following sessions:
``` 
            { 
                SubjectId = "sub1",
                SessionId = "sid1_1",
            }
            { 
                SubjectId = "sub1",
                SessionId = "sid1_2",
            }, 
```
sub1 will point to the newest cache entry
sid_1_1 to the old cache entry
sid_1_2 to the new cache entry

File: src/Duende.Bff.DistributedCache/UserSessionCacheExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            Identifier identifier = session.GetKeyIdentifier(applicationDiscriminator);
+            Identifier subjectId = session.GetSubjectIdentifier(applicationDiscriminator);
+            Identifier sessionId = session.GetSessionIdentifier(applicationDiscriminator);
+
+            string? serializedSession = session.ToJson();
+
+            DistributedCacheEntryOptions cacheOptions = new DistributedCacheEntryOptions
+            {
+                AbsoluteExpiration = session.Expires ?? DateTimeOffset.Now.AddDays(1)
```
Comment (PascalSenn) [inline src/Duende.Bff.DistributedCache/UserSessionCacheExtensions.cs:27]: Not sure if this makes sense
The default expiry of the cache entry are the expiry dates of the session.
If nothing is set it is by default one day

File: test/Duende.Bff.DistributedCache.Tests/Duende.Bff.DistributedCache.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (PascalSenn) [inline test/Duende.Bff.DistributedCache.Tests/Duende.Bff.DistributedCache.Tests.csproj:1]: Some of these tests will fail atm, as of the limitations

**Comment by brockallen** [comment]
Hi, thanks for this PR -- I finally had time to have a look. It's an interesting approach -- basically keeping multiple entries for the one logical entry all to support the various queries.

You had a comment about one entry only being able to refer to the latest... perhaps another dedicated entry that keeps a list of the other ids? Not sure what extra concurrency bugs that might introduce, though.

**Comment by joseftw** [comment]
Any updates on this? I would be happy to help getting this feature merged. Right now we are using a custom implementation, would be awesome if this was provided ootb. :)

**Comment by brockallen** [comment]
I did some more thinking. I'm wondering if we can have these cache entries with the following prefixes: 

key:key, value:(sub,sid) => this will get you to the real item
sub:sub, value:[sid, sid, sid] => this will give you a list of the sids for the current user, which gets you to all the items
sid:sid, value:sub => for a sid, this will get you to the real item
item:(sub,sid), value: the real payload, which contains the key, sub, and sid used to cleanup the other entries

For example, with a sub of 123, sid of 456 and 789, and keys 888 and 999 (for the 2 sessions) we might have these entries:

key:888, value: (123, 456)
key:999, value: (123, 789)
sub:123, value: [456, 789]
sid:456, value: 123
sid:789, value: 123
item:(123,456): value: {claims, props, sub:123, sid:456, key:888}
item:(123,789): value: {claims, props, sub:123, sid:789, key:999}

So then the APIs on the store are:
* Get(key) => look up "key:888", then use that to find "item:(123,456)"
* Update(key, session) => look up "key:888", then use that to find "item:(123,456)" and update all associated entries to extend cache items
* Delete(key) => look up "key:888", then use that to find "item:(123,456)" and delete all entries. for the "sub:123" entry the array would be adjusted and the item updated in the cache
* GetSessions(sub?, sid?) => use what you have (sub and/or sid) to ultimately get to the right entries
* DeleteSessions(sub?, sid?) => use what you have (sub and/or sid) to ultimately delete the right entries

Basically we're re-inventing a DB index, I guess. Thoughts?

**Comment by PascalSenn** [comment]
@brockallen  We probably still have concurrency issues, but i guess these would be resloved over time. I can update the PR

**Comment by brockallen** [comment]
>  We probably still have concurrency issues, but i guess these would be resloved over time. I can update the PR

Sure. We can always try to minimize those by taking a semantic lock on the sid (or something) so at least a single server doesn't step on itself. Also, if a browser is making many concurrent requests then there's always the browser as a concurrency issue, meaning 2 concurrent http responses updating the cookie -- which one wins?

**Comment by brockallen** [comment]
Also, the mainline has updated quite a bit, so you might need to rebase.

**Comment by brockallen** [comment]
Hi @PascalSenn. I wanted to follow up on this PR. What are you thinking? Is this something you still would like to pursue, or not?

**Comment by PascalSenn** [comment]
@brockallen  sorry completely forgot about this PR. Yes, i will tidy it up next week 👍

**Comment by brockallen** [comment]
Ok, great! Also, we will need to rebase to our 2.0 version.... that might be easier to just start a new branch/PR.

**Comment by PascalSenn** [comment]
@brockallen  Sorry I do not have the time to finish this off in a reasonable timeframe. I will pick it up when i get the time.

**Comment by brockallen** [comment]
Understood. As we have discovered, this is not a trivial task.
